tensorflow_reader/tensorflow_reader.py

```diff
@@ -95,6 +97,622 @@ def decode(self, buf, out=None):
+            width, height = os_obj.level_dimensions[level]
+
+        elif re.compile(r"\.zarr$").search(filename):
+            # read whole-slide image and create zarr objects
+            store = zarr.DirectoryStore(filename)
+            source_group = zarr.open(store, mode="r")
+
+            # measure objective of level 0
+            objective = np.float32(source_group.attrs[os.PROPERTY_NAME_OBJECTIVE_POWER])
+
+            # calculate magnifications of levels
+            estimated = np.array(objective / source_group.attrs["level_downsamples"])
+
+            # Find best level to use and its factor
+            level, factor = self._get_level_and_factor(magnification, estimated, tolerance)
+
+            # get slide width, height at desired magnification. (Note height before width)
+            height, width = source_group[format(level)].shape[0:2]
+
+        else:
+            # We don't know magnifications so assume reasonable values for level and factor.
+            level = 0
+            factor = 1.0
+            if True:
+                pil_obj = Image.open(filename)
+                width, height = pil_obj.size
+            else:
+                # For the case that we know the image size without opening the file
+                width, height = 2048, 2048
+
+        print(f"level = {level}, factor = {factor}, width = {width}, height = {height}")
+        return level, factor, width, height
+
+    def _get_level_and_factor(self, magnification, estimated, tolerance):
+        # calculate difference with magnification levels
+        delta = magnification - estimated
+
+        # match to existing levels
+        if np.min(np.abs(np.divide(delta, magnification))) < tolerance:  # match
+            level = np.squeeze(np.argmin(np.abs(delta)))
+            factor = 1.0
+        elif np.any(delta < 0):
+            value = np.max(delta[delta < 0])
+            level = np.squeeze(np.argwhere(delta == value)[0])
+            factor = magnification / estimated[level]
+        else:  # desired magnification above base level - throw error
+            raise ValueError("Cannot interpolate above scan magnification.")
+
+        return level, factor
+
+
+"""
+
+AddTileDescription: An instance of class AddTileDescription can be supplied as an argument to
+`tensorflow.dataset.map`.  AddTileDescription adds new key-value pairs to the tensorflow dictionary to
+set the desired tile width, height, width overlap, and height overlap for each element, and to indicate
+whether we want tiles even if they are fractional (aka of truncated size) due to partially falling off
+the edge of the image.  (These fractional tiles can be problematic because tensorflow likes its shapes
+to be uniform.)  chunk_width_factor and chunk_height_factor indicate how many tiles are read at a time.
+The primary functionality of this class over an ordinary dictionary is that it sets all the required
+keys and no others.
+
+"""
```

Comment:
We want these descriptions to go after the class declaration so they are used as docstrings.

Reply:
Thank you.  Pushed.